doc/design/wallet/wallet-design.md

```diff
@@ -0,0 +1,762 @@
+                                           ///
+                                           /// #Params
+                                           /// storage_handle: opened storage handle (See open handler)
+                                           /// type_: allows to separate different record types collections
+                                           /// id: the id of record
+                                           /// tag_names_json: the list of tag names to remove from the record as json array:
+                                           ///   ["tagName1", "tagName2", ...]
+                                           ///   Note that null means no tag names
+                                           delete_record_tags: Option<extern fn(storage_handle: i32,
+                                                                                type_: *const c_char,
+                                                                                id: *const c_char,
+                                                                                tag_names_json: *const c_char) -> ErrorCode>,
+
+                                           /// Delete an existing record in the wallet storage
+                                           ///
+                                           /// #Params
+                                           /// storage_handle: opened storage handle (See open handler)
+                                           /// type_: record type
+                                           /// id: the id of record
+                                           delete_record: Option<extern fn(storage_handle: i32,
+                                                                           type_: *const c_char,
+                                                                           id: *const c_char) -> ErrorCode>,
+
+                                           /// Get an wallet storage record by id
+                                           ///
+                                           /// #Params
+                                           /// storage_handle: opened storage handle (See open handler)
+                                           /// type_: allows to separate different record types collections
+                                           /// id: the id of record
+                                           /// options_json: //TODO: FIXME: Think about replacing by bitmaks
+                                           ///  {
+                                           ///    retrieveValue: (optional, true by default) Retrieve record value,
+                                           ///    retrieveTags: (optional, true by default) Retrieve record tags
+                                           ///  }
+                                           /// record_handle_p: pointer to store retrieved record handle
+                                           get_record: Option<extern fn(storage_handle: u32,
```

Comment:
I think we do not need functions with record_handle. In get_record we can return the data immediately (if requested) or set the returned pointer to NULL if not requested.
I had in mind function like this: 
```
get_record: Option<extern fn(handle: i32,
                                class_p: *const c_char,
                                class_len: c_int,
                                name_p: *const c_char,
                                name_len: c_int,
                                options: *const c_char,
                                value_p: *mut *const c_char,
                                value_len: *mut c_int,
                                value_key_p: *mut *const c_char,
                                value_key_len: *mut c_int,
                                tags_p: *mut *const c_char) -> ErrorCode>;
```

Reply:
I have big concern about this proposal. When you provide the pointer you transfer memory ownership and should provide the way to free memory. We have mix of types here and will require a lot of handlers to free memory. Providing record_handle solve this problem. We have just one free_record function that consumes this handle.

Reply:
OK, this is acceptable.